### PR TITLE
Soft-delete attribute type choice options

### PR DESCRIPTION
**Pull request by bbliem**
# Archive `AttributeTypeChoiceOption` instead of deleting when references exist

## Why

Hard-deleting an `AttributeTypeChoiceOption` cascades to its referencing
`AttributeChoice` / `AttributeChoiceWithText` rows, but does **not** clean up
historical references stored as PKs in:

- `wagtail.models.Revision.content` JSON (drafts and history of `Action`),
- `reversion.models.Version.serialized_data` of `AttributeChoice` /
  `AttributeChoiceWithText` (used by report snapshots).

When such a stale reference is later dereferenced (e.g. via
`ActionSnapshot.get_attribute_for_type` while rendering an action's report
tab), the AttributeChoice's `__str__` raises `ObjectDoesNotExist` and is shown
as `"Missing value"`. We observed this on Action 13074 under report 108 for
the deleted `AttributeTypeChoiceOption` 807.

This PR introduces a lifecycle for choice options so that customers can retire
options without breaking historical data, while still allowing genuine
hard-delete when no references exist.

### Model lifecycle (commit 1)

- New `is_active` BooleanField on `AttributeTypeChoiceOption`, default `True`.
- New manager / queryset helpers `objects.active()` and `objects.archived()`.
- `archive()` and `unarchive()` methods. Both update `order` to keep the
  uniqueness constraint `(type, order)` satisfied as rows shuttle between the
  active and archived ranges.
- `is_referenced()` returns `True` when the option is pinned by:
  - a live `AttributeChoice` / `AttributeChoiceWithText` row, or
  - a Wagtail `Revision.content` referencing the option's PK, or
  - a reversion `Version` of `AttributeChoice` / `AttributeChoiceWithText`
    referencing the PK.

  The query dispatches on `AttributeType.format` so only the relevant shape
  runs: one live FK check (not both), one JSONB structural containment
  against `Revision.content` (`content @> '{"attributes": {<fmt>: ...}}'`),
  and one `content_type__model=` filter on `reversion.Version` plus
  `,` / `}`-anchored substring patterns to avoid substring false positives
  (e.g. `"choice": 15020` matching PK 1502), followed by a per-row JSON
  parse to confirm. Options on a non-choice `AttributeType` short-circuit
  to `False`.

Live-value policy: archiving an option leaves the corresponding live
`AttributeChoice` rows in place. Existing actions continue to display the
option's value; only new selections are blocked.

### Picker visibility and write-side rejection (commit 2)

Read side — keep archived options resolvable but out of selection UIs:

- The action editor's `OrderedChoice` and `OptionalChoiceWithText` form
  pickers filter the choice queryset to active options. If the current
  attribute value happens to point at an archived option, that option is
  kept in the dropdown for that form so existing data isn't silently
  dropped on the next save.
- GraphQL: `AttributeTypeNode.choiceOptions` lists **all** options
  (active + archived) with an `isActive` field on each. REST consumers
  (e.g. the spreadsheet table editor) that receive raw choice PKs in
  action data depend on this to resolve the label regardless of archive
  state; clients filter to `isActive=true` when building selection UIs.
- FK dereferencing is **not** filtered — historical revisions, report
  snapshots, and existing live values all keep resolving to the option's
  name.

Write side — reject archived options as new selections, with the rule
enforced at every layer that admits a new value:

- The Strawberry `updateAction` / `createAction` mutation validates the
  incoming `choice_id` in `_set_action_choice_attribute` and rejects
  archived options unless the action already has that exact option as
  its current value.
- The DRF REST serializers' `ChoiceAttributesSerializer.validate()` /
  `ChoiceWithTextAttributesSerializer.validate()` run the same rejection
  during DRF's `is_valid()` phase, returning a 400 immediately and before
  any save side effect happens. (The bulk REST path issues
  `QuerySet.bulk_update()` / `bulk_create()`, which bypass both `save()`
  and `full_clean()`, so the model-layer guards alone are insufficient
  *and* would only fire after the Action save had already run.)
- `ModelWithAttributes.set_attribute()` runs the rule before queuing the
  create/update operation, covering any code that goes through the
  attribute-serialization machinery.
- `AttributeChoice` and `AttributeChoiceWithText` enforce the same rule
  from `save()` and `clean()`, covering anything that bypasses the
  serializer layer and goes through Django's standard model machinery.

The shared validator accepts the choice in three cases:

- when it's `None` or active,
- when the row already has that exact option as its current value
  (in-place re-save of an unchanged archived value), or
- when a Wagtail revision of the parent already references this
  `(type, choice)` pair. This covers the publish-from-draft path:
  `Action.commit_attributes()` calls `AttributeType.create_attribute()`
  which saves a brand-new `AttributeChoice` (`pk=None`) carrying the
  archived option that was picked in the draft. Without this exemption
  the model-layer guard would block users from publishing drafts that
  pre-date the archive — even though `is_referenced()` deliberately
  kept the option around because of that very draft.

The target-instance lookup used by the REST validator walks the parent
chain so the same code path resolves the row whether it lives on a
single-instance request (outer serializer's `instance`) or under a
bulk wrapper (per-row `_instance` set by
`BulkSerializerValidationInstanceMixin` during each child's
`run_validation`). The earlier `hasattr(parent, 'objs_by_id'): return
None` bail-out would have lost track of the row whenever the bulk
wrapper sat between the choice serializer and the per-row child,
forcing `existing_by_identifier` to stay empty and rejecting every
archived option in the payload — including the action's own current
value.

### Archive-on-delete in the inline editor (commit 3)

- New `ArchivableOrderedModelChildFormSet` in `aplans/utils.py`. Subclasses
  the existing `OrderedModelChildFormSet`; overrides `delete_existing` to
  call `obj.archive()` when `obj.is_referenced()` returns `True`, and pulls
  the object out of `self.deleted_objects` so modelcluster's commit doesn't
  strip it from the database.
- Wired in as the `choice_options` formset on `AttributeTypeAdmin`.
- New POST-only admin URL `/admin/actions/attributetype/<pk>/unarchive-choice-option/<option_pk>/`
  guarded by the AttributeType edit permission, calling `option.unarchive()`.

Choice options that are not referenced anywhere still hard-delete as before.

### Inline editor surface (commit 4)

- New `ChoiceOptionArchivePanel` rendered at the top of an expanded
  archived option in the inline editor: a Wagtail "Archived" status badge,
  a one-line explanation, and an Unarchive button that POSTs to the
  unarchive URL via `fetch`. Active rows render nothing.
- The button uses inline JS rather than a nested `<form>` so it doesn't
  interfere with the surrounding Wagtail edit form.

### Surfacing what happened to the admin (commit 5)

- The expanded form's existing `ChoiceOptionUsagePanel` is reworded to
  explain up-front that deletion will become archive when the option is
  in use, and points at the unarchive button as the recovery path.
- `ArchivableOrderedModelChildFormSet.save()` records archived items on
  `self.archived_on_last_save`. `AttributeTypeEditView.form_valid()`
  reads each formset's `archived_on_last_save` after the parent save and
  emits one `messages.warning` per archived option, naming it.

### Preserve position across archive/unarchive (commit 6)

Originally `archive()` bumped `order` to `max+1` so the inline editor's
compaction wouldn't collide with archived rows under
`unique_order_per_type`. That made archive→unarchive lose the row's
original position — it always reappeared at the end of the list.

`archive()` and `unarchive()` now just flip `is_active`. The harder part
is the inline-formset save: modelcluster always renumbers active rows
to `0..N-1` in formset display order, which leaves a freshly-archived
row's `order` (untouched) *higher* than the now-compacted active
siblings — so the archived row still appears at the end of the sort
even though its `order` field wasn't changed.

`ArchivableOrderedModelChildFormSet.save()` now snapshots every row's
`order` *before* calling the parent save and then runs a single
collision-safe two-phase `UPDATE` afterwards. When at least one row
was archived this save, every surviving row (active *and* archived)
is restored to its pre-save `order` so the archived row keeps its slot
in the sort. On saves that didn't archive anything, active rows are
compacted to formset display order — the baseline
`OrderedModelChildFormSet` behaviour.

The two-phase update first parks every affected row at a high disjoint
slot (`shift + final_target`, where `shift = max(existing, target) + 1`),
then brings them down to their final positions. Within each phase the
slots are unique by construction, so no single statement can violate
the `unique (parent, order)` constraint — even though it's
`deferrable=DEFERRED`, this avoids relying on the surrounding code
path holding a transaction open for the deferred check to ride on.

Trade-off: an admin reorder performed in the *same* save as an archive
gets undone (the restore puts active rows back at their pre-save
positions). Admins rarely combine the two, and the simpler reconciliation
is worth the corner case.

## Tests

In `actions/tests/test_attribute_type_choice_option_deletion.py`:

- `TestArchiveApi` — `archive` / `unarchive` semantics, queryset filters,
  `is_referenced` across live / Wagtail-revision / reversion-Version
  sources, "false when unused", ordering doesn't collide between active
  and archived rows.
- `TestPickerVisibility` — ordered- and optional-choice pickers exclude
  archived options; an existing value pointing at an archived option is
  preserved; GraphQL listing includes archived options and tags each one
  with `isActive`; model-level rejection via `full_clean()`, `save()`,
  and `set_attribute()` (covering creates, in-place re-saves, and
  switches to archived); REST integration test against
  `PUT /v1/plan/<id>/actions/` confirming an archived assignment returns
  400 and writes nothing.
- `TestArchivableFormset` — inline formset archives referenced options on
  delete, hard-deletes unreferenced ones, records archived items on
  `archived_on_last_save`.
- `TestUnarchiveView` — POST flips the flag, GET is 405, anonymous gets
  302/403, wrong-attribute-type returns 404.
- `TestArchivePanel` — `ChoiceOptionArchivePanel` shown only for archived
  options, renders badge + unarchive URL, hidden for unsaved options.
- `TestArchiveNotification` — `AttributeTypeEditView.form_valid` emits
  one warning per archived option across all formsets; emits nothing
  when nothing was archived.
- `TestUsagePanelWording` — usage panel renders the new "deletion
  becomes archive" wording for in-use options.

In `actions/tests/test_graphql_mutations.py::TestUpdateAction`:

- `test_update_action_rejects_archived_choice_as_new_value` — submitting
  an archived option as a new value returns a `VALIDATION` OperationInfo
  message and writes no AttributeChoice.
- `test_update_action_allows_resaving_existing_archived_choice` — an
  action whose value is an option that has since been archived can be
  re-saved with that same option (no-op selection) without rejection.
- `test_update_action_rejects_switching_to_archived_choice` — when the
  existing value is an active option, switching to a different,
  archived option is rejected and the existing value remains intact.

Total: 61 archive-specific tests in `test_attribute_type_choice_option_deletion.py`
and 3 in `test_graphql_mutations.py::TestUpdateAction` (123 tests pass
across the two files together). A broader sweep of `actions/tests/`
shows no regressions (one pre-existing staticfiles-manifest failure in
`test_can_access_plan_edit_page` is unrelated to this branch).

## Migration

- `actions/0175_attributetypechoiceoption_is_active.py` — adds the
  `is_active` column with `default=True`, so all existing rows become
  active on apply. Forward-only schema migration, no data backfill needed.

## Operator follow-up: repairing the original incident

This PR provides the mechanism but does not by itself restore missing
`AttributeTypeChoiceOption` instances (e.g., the one with PK 807). To fix the
original "Missing value" symptom (e.g., on Action 13074), an operator needs to
re-create such choice options with `is_active=False`, using their
name/identifier from a DB backup or from reversion's own `Version` history of
`AttributeTypeChoiceOption`. The mechanism added here will then keep that
resurrected row out of new pickers while letting the historical FK dereferences
resolve to the original name.

## Follow-up PR

We should make improvements to the table editor to omit archived choice options from dropdowns (unless they are currently selected).

## Related issue
[Sentry issue](https://sentry.kausal.tech/organizations/kausal/issues/13659/events/5b001f2dc1f441cd959ed89aacaf50a9/)

------

### Type of Change
- [x] Set the PR's label to match the nature of this change

### Testing
- [x] **Built Unit tests** (unit tests added/updated)
- [ ] **Built E2E tests** (if applicable. E2E tests added/updated)
- [ ] **Authorization is tested** (permissions and access controls verified)
- [x] **Manually tested locally** (functionality verified)
    ##### Manual testing instructions
    If feature requires manual testing by reviewer, you can provide instructions here.

### Internationalization & Accessibility
- [x] **New strings are translatable** (all user-facing text uses i18n)
- [ ] **Accessibility** standards met (WCAG compliance, screen reader support)

### Integrations (if applicable)
If there are model changes to models which use any of the features below, verify the new models work together with the features.
For example, when adding a new model, verify the new model instances are copied when copying a plan.
- [ ] **Moderation** integration implemented
- [ ] **Reporting** integration implemented
- [ ] **Copy plan feature** integration implemented
- [ ] **Umbrella plan structure** integration implemented

### Dependencies
- [ ] **Dependencies are merged** (if applicable. If the change depends on other PRs e.g. kausal_common)

### CODE REVIEW

File: aplans/utils.py [hunk 1]
```diff
@@ -290,6 +291,47 @@ def save(self, commit=True):
...
+            # Pull this object out of `deleted_objects` so modelcluster's
+            # `manager.remove(*self.deleted_objects)` doesn't strip it from
+            # the cluster — which would then delete it on commit().
+            with suppress(ValueError):
+                self.deleted_objects.remove(obj)
+            return
+        super().delete_existing(obj, commit=commit)
+
+    def save(self, commit=True):
+        saved_instances = super().save(commit)
```
[the remark this answers is not in the document]
Comment (bbliem) [reply aplans/utils.py]: No longer applies.

File: actions/api.py [hunk 2]
```diff
@@ -604,6 +604,70 @@ def to_value_parameter(self, item: Any) -> dict[str, Any]:
...
+    """Find the model instance whose existing values to compare against."""
+    parent = serializer.parent
+    if parent is None:
+        return None
+    target = getattr(parent, '_instance', None)
+    if target is not None:
+        return target
+    if hasattr(parent, 'objs_by_id'):
+        # Bulk wrapper without `_instance` set yet — no per-instance target.
+        return None
```
[the remark this answers is not in the document]
Comment (bbliem) [reply actions/api.py]: Fixed.

File: actions/models/attributes.py [hunk 3]
```diff
@@ -362,6 +500,41 @@ def __str__(self):
...
+    is_active = AttributeTypeChoiceOption.objects.filter(pk=attribute.choice_id).values_list('is_active', flat=True).first()
+    if is_active is None or is_active:
+        # Either the option doesn't exist (let FK validation handle it) or it
+        # is still active — nothing to do.
+        return
+    previous_choice_id = (
+        type(attribute).objects.filter(pk=attribute.pk).values_list('choice_id', flat=True).first()
+        if attribute.pk is not None
+        else None
+    )
```
[the remark this answers is not in the document]
Comment (bbliem) [reply actions/models/attributes.py]: Fixed.

File: aplans/utils.py [hunk 4]
```diff
@@ -290,6 +291,86 @@ def save(self, commit=True):
...
+        restores the option to its original DB slot with no further work.
+        """
+        archived_orders = set(
+            self.model._default_manager.filter(
+                **{self.fk.name: self.instance},
+                is_active=False,
+            ).values_list('order', flat=True),
+        )
+        next_order = 0
+        for form in self.ordered_forms:
```
[the remark this answers is not in the document]
Comment (bbliem) [reply aplans/utils.py]: No longer applies.

File: aplans/utils.py [hunk 5]
```diff
@@ -290,6 +291,127 @@ def save(self, commit=True):
...
+            current_pks = list(
+                self.model._default_manager.filter(**parent_filter).values_list(
+                    'pk', flat=True,
+                ),
+            )
+            max_pre = max(pre_save_orders.values(), default=-1)
+            next_new = max_pre + 1
+            for pk in current_pks:
+                if pk in pre_save_orders:
+                    targets[pk] = pre_save_orders[pk]
```
[the remark this answers is not in the document]
Comment (bbliem) [reply aplans/utils.py]: No longer applies.

**Comment by tituomin** [comment]
UX issue: I don't think it's optimal to dedicate so many warning messages to the delete case even when most of the time the user is not even considering deleting. Could we put this into a toast message or conditionally display only when the trash icon is pressed?
image

**Comment by tituomin** [comment]
Word choices:

>  existing references continue to resolve unchanged

I feel this is correct but perhaps too succinct / abstract for users.

How about something like: it will still be visible wherever it has already been selected as a value.

**Review by tituomin — COMMENTED** [review]
Looks good! I have some usability / wording comments.
My biggest comment though is that I'm not sure if it's worth some of the extra code to try to preserve the ordering of archived -> unarchived choices. It should be trivial to reorder the choices by hand after unarchiving, shouldn't it?

**Comment by bbliem** [comment]
> Looks good! I have some usability / wording comments. My biggest comment though is that I'm not sure if it's worth some of the extra code to try to preserve the ordering of archived -> unarchived choices. It should be trivial to reorder the choices by hand after unarchiving, shouldn't it?

@tituomin Thanks for the feedback. I thought about this, and in fact the implementation originally made for this PR did not preserve the ordering. Preserving the order was added in commit 835fe200a. We could revert it, but I implemented order preservation because, in general, the order is meaningful and important for customers. Losing the order by archiving / unarchiving is an unpleasant surprise. It might also not be noticed. We can discuss this further.

File: actions/models/attributes.py [hunk 6]
```diff
@@ -340,9 +370,105 @@ def __str__(self):
...
+            )
+            .filter(
+                Q(serialized_data__contains=f'"type": {type_id},') | Q(serialized_data__contains=f'"type": {type_id}}}'),
+            )
+        )
+
+        for v in candidates.iterator():
+            try:
+                payload = json.loads(v.serialized_data)
+            except TypeError, ValueError:
```
[the remark this answers is not in the document]
Comment (bbliem) [reply actions/models/attributes.py:478]: False alarm. This may be true for Python 2, but not for Python 3.

File: aplans/utils.py [hunk 7]
```diff
@@ -290,6 +291,64 @@ def save(self, commit=True):
...
+        super().delete_existing(obj, commit=commit)
+
+    def save(self, commit=True):
+        # `delete_existing()` is called from inside `super().save()` (with
+        # `commit=False`, regardless of the outer commit flag). Track the
+        # outer commit here so the delete hook can decide whether to write
+        # the archive immediately.
+        self._outer_commit = commit
+        try:
+            saved_instances = super().save(commit)
```
[the remark this answers is not in the document]
Comment (bbliem) [reply aplans/utils.py:343]: This doesn't appear to be an issue in production. The archived-row drag can only raise IntegrityError when the save runs outside a request transaction; under `ATOMIC_REQUESTS=True` the whole admin save is one transaction and `unique_order_per_type` is DEFERRED, so the renumber's transient duplicates are tolerated and the distinct final ordering commits cleanly.
